cortex-a-rt/src/lib.rs

```diff
@@ -376,15 +417,65 @@ macro_rules! restore_context {
 // Our assembly language exception handlers
 core::arch::global_asm!(
     r#"
-    .section .text.handlers
-    .align 0
+    .section .text._asm_default_undefined_handler
+
+    // Called from the vector table when we have an undefined exception.
+    // Saves state and calls a C-compatible handler like
+    // `extern "C" fn _undefined_handler(addr: usize) -> usize;`
+    // or
+    // `extern "C" fn _undefined_handler(addr: usize) -> !;`
+    .global _asm_default_undefined_handler
+    .type _asm_default_undefined_handler, %function
+    _asm_default_undefined_handler:
+        // state save from compiled code
+        srsfd   sp!, {und_mode}
+        // to work out what mode we're in, we need R0
+        push    {{r0}}
+        // First adjust LR for two purposes: Passing the faulting instruction to the C handler,
+        // and to return to the failing instruction after the C handler returns.
+        // Load processor status for the calling code
+        mrs     r0, spsr
+        // Was the code that triggered the exception in Thumb state?
+        tst     r0, {t_bit}
+        // Subtract 2 in Thumb Mode, 4 in Arm Mode - see p.1206 of the ARMv7-A architecture manual.
+        ite     eq
+        subeq   lr, lr, #4
+        subne   lr, lr, #2
+        // save the newly computed LR
+        push    {{lr}}
+        // now do our standard exception save
+    "#,
+    save_context!(),
+    r#"
+        // Pass the faulting instruction address to the handler.
+        mov     r0, lr
+        // call C handler
+        bl      _undefined_handler
+        // if we get back here, assume they returned a new LR in r0
+        mov     lr, r0
+        // do our standard restore
+    "#,
+    restore_context!(),
+    r#"
+        // get our saved LR
+        pop     {{lr}}
```

Comment:
This will clobber the `lr` that was returned by `_undefined_handler`, right? The advertised API suggests the caller can cause the handler to return to an arbitrary new instruction, and `mov lr, r0` and associated comment appear to be trying to support that, but that register move is then discarded here and replaced with the value that was pushed before `save_context`.

Reply:
See https://github.com/rust-embedded/cortex-ar/issues/29